authentication/oidc.go

```diff
@@ -45,71 +47,127 @@ type OIDCConfig struct {
 // Middleware is a convenience type for functions that wrap http.Handlers.
 type Middleware func(http.Handler) http.Handler
 
-// NewOIDC creates a single http.Handler and a set of Middlewares for all
+// OIDCHandlers holds handlers and a set of middlewares for all
 // tenants that is able to authenticate requests and provide the
 // authorization code grant flow for users.
-func NewOIDC(
-	logger log.Logger,
-	reg prometheus.Registerer,
-	prefix string,
-	configs []TenantOIDCConfig,
-) (http.Handler, map[string]Middleware, []error) {
-	handlers := map[string]http.Handler{}
-	middlewares := map[string]Middleware{}
-	warnings := make([]error, 0, len(configs))
-
-	tenantErrors := prometheus.NewGauge(prometheus.GaugeOpts{
+type OIDCHandlers struct {
+	mtx         sync.RWMutex
+	handlers    map[string]http.Handler
+	middlewares map[string]Middleware
+
+	logger     log.Logger
+	retryGauge prometheus.Gauge
+}
+
+// NewOIDCHandlers instantiates OIDC handlers.
+func NewOIDCHandlers(l log.Logger, reg prometheus.Registerer) *OIDCHandlers {
+	tenantsFailing := prometheus.NewGauge(prometheus.GaugeOpts{
 		Namespace: "observatorium",
 		Subsystem: "api",
-		Name:      "tenant_registration_failures",
-		Help:      "The number of tenants for which OIDC provider instantiation failed.",
+		Name:      "tenants_failing_registrations",
+		Help:      "The number of tenants for which OIDC provider instantiation is failing.",
 	})
 
-	reg.MustRegister(tenantErrors)
+	reg.MustRegister(tenantsFailing)
 
-	for _, c := range configs {
-		p, err := NewProvider(context.TODO(), logger, getCookieForTenant(c.Tenant), "/"+c.Tenant, c.OIDCConfig)
-		if err != nil {
-			warnings = append(warnings, fmt.Errorf("failed to instantiate OIDC provider for tenant %q: %w", c.Tenant, err))
+	return &OIDCHandlers{
+		handlers:    make(map[string]http.Handler),
+		middlewares: make(map[string]Middleware),
+		logger:      l,
+		retryGauge:  tenantsFailing,
+	}
+}
 
-			tenantErrors.Inc()
+// AddOIDCForTenant adds a new handler and middleware for a new tenant, based on
+// the tenant OIDC configuration. If a connection with the OIDC provider
+// cannot be established, the method will retry to connect with an increasing
+// backoff.
+func (oh *OIDCHandlers) AddOIDCForTenant(prefix string, config TenantOIDCConfig) {
+	const (
+		loginRoute    = "/login"
+		callbackRoute = "/callback"
+	)
+
+	ctx := context.Background()
```

Comment:
Since you spawn go routines I would ensure:
* We get this context hooked to interrupt handler (so when we do ctrl+c we clean go routine nicely, not aborting in random moment) - good practice
* Maybe TODO, but if we want to put this handler on HTTP then what we do if I add OIDC tenant `abc` and it's retrying in loop and  another client adds `abc` and 100 more? Do we need need to handle this somehow?

Reply:
For your second point, do you mean if there are duplicate tenants trying to register or if there is a large number of tenants running the registration loop at the same time?

Reply:
Duplicate tenants

Reply:
Hm I guess then we should have a means of detecting duplicate tenants in config, possibly before we proceed to further steps. If that's fine I'd leave it as a TODO, but I agree this should be taken care of.

Reply:
👍🏽